File: Week 10/rust/1.matrix_basics.rs
/*
 * WEEK 10 - RUST DSA
 * Topic: Matrix Basics
 * File: 1.matrix_basics.rs
 *
 * CONCEPT:
 *     A 2D matrix as Vec<Vec<i32>>. Supports transpose (square in-place
 *     and general), row reverse, 90° CW/CCW rotation, naive multiply.
 *
 * KEY POINTS:
 *     - In-place transpose only works for square matrices.
 *     - 90° CW = transpose + reverse each row.
 *     - 90° CCW = reverse each row + transpose.
 *
 * ALGORITHM / APPROACH:
 *     See per-function code; same as Java/Python.
 *
 * RUST-SPECIFIC NOTES:
 *     - Use Vec<Vec<i32>> for the row-major matrix; a flat Vec<i32> with
 *       manual indexing is the cache-friendlier alternative.
 *     - .reverse() on a slice mutates in place.
 *     - For high-performance numeric work use the ndarray or nalgebra crates.
 *
 * DRY RUN:
 *     [[1,2,3],[4,5,6],[7,8,9]]
 *         transpose: [[1,4,7],[2,5,8],[3,6,9]]
 *         rotate CW: [[7,4,1],[8,5,2],[9,6,3]]
 *     [[1,2],[3,4]] x [[5,6],[7,8]] = [[19,22],[43,50]]
 *
 * COMPLEXITY:
 *     Transpose      : O(rows*cols)
 *     Rotation       : O(n^2)
 *     Multiplication : O(m*n*k)
 */

type Matrix = Vec<Vec<i32>>;

fn print_matrix(mat: &Matrix) {
    for row in mat {
        println!("{:?}", row);
    }
    println!();
}

fn transpose_square(mat: &mut Matrix) {
    let n = mat.len();
    for i in 0..n {
        for j in (i + 1)..n {
            let tmp = mat[i][j];
            mat[i][j] = mat[j][i];
            mat[j][i] = tmp;
        }
    }
}

fn transpose(mat: &Matrix) -> Matrix {
    let rows = mat.len();
    let cols = mat[0].len();
    let mut result: Matrix = vec![vec![0; rows]; cols];
    for i in 0..rows {
        for j in 0..cols {
            result[j][i] = mat[i][j];
        }
    }
    result
}

fn reverse_rows(mat: &mut Matrix) {
    for row in mat.iter_mut() {
        row.reverse();
    }
}

fn rotate_90_cw(mat: &mut Matrix) {
    transpose_square(mat);
    reverse_rows(mat);
}

fn rotate_90_ccw(mat: &mut Matrix) {
    reverse_rows(mat);
    transpose_square(mat);
}

fn multiply(a: &Matrix, b: &Matrix) -> Matrix {
    let m = a.len();
    let k = a[0].len();
    let n = b[0].len();
    let mut c: Matrix = vec![vec![0; n]; m];
    for i in 0..m {
        for j in 0..n {
            let mut sum = 0;
            for p in 0..k {
                sum += a[i][p] * b[p][j];
            }
            c[i][j] = sum;
        }
    }
    c
}

fn main() {
    let mat: Matrix = vec![
        vec![1, 2, 3],
        vec![4, 5, 6],
        vec![7, 8, 9],
    ];
    println!("Original:");
    print_matrix(&mat);

    let mut mat2 = mat.clone();
    transpose_square(&mut mat2);
    println!("Transposed (square, in-place):");
    print_matrix(&mat2);

    let non_square: Matrix = vec![vec![1, 2, 3], vec![4, 5, 6]];
    println!("Non-square (2x3):");
    print_matrix(&non_square);
    println!("Transposed (3x2):");
    print_matrix(&transpose(&non_square));

    let mut mat3 = mat.clone();
    rotate_90_cw(&mut mat3);
    println!("Rotated 90° CW:");
    print_matrix(&mat3);

    let a: Matrix = vec![vec![1, 2], vec![3, 4]];
    let b: Matrix = vec![vec![5, 6], vec![7, 8]];
    println!("A x B:");
    print_matrix(&multiply(&a, &b));

    // Rotate CCW demo
    let mut mat4 = mat.clone();
    rotate_90_ccw(&mut mat4);
    println!("Rotated 90° CCW:");
    print_matrix(&mat4);
}

/*
 * NOTES — Rust vs Java:
 *     - vec![vec![0; cols]; rows] avoids the aliasing trap that
 *       Python's [[0]*cols]*rows hits.
 *     - .reverse() on a slice/Vec mutates in place.
 *     - For real numeric work prefer the ndarray crate (cache-friendly,
 *       BLAS-backed multiplications).
 */

File: Week 10/cpp/1.matrix_basics.cpp
/*
 * WEEK 10 - C++ DSA
 * Topic: Matrix Basics
 * File: 1.matrix_basics.cpp
 *
 * CONCEPT:
 *     Use std::vector<std::vector<int>> as a 2D matrix.
 *     Operations: print, transpose (square in-place + general),
 *     row reverse, 90° CW/CCW rotation, naive matrix multiplication.
 *
 * KEY POINTS:
 *     - In-place transpose only works on square matrices.
 *     - 90° CW = transpose + reverse each row.
 *     - 90° CCW = reverse each row + transpose.
 *     - Naive matmul is O(m * n * k).
 *
 * ALGORITHM / APPROACH:
 *     See per-function code; same as Java/Python.
 *
 * C++-SPECIFIC NOTES:
 *     - std::reverse(row.begin(), row.end()) reverses a row.
 *     - Pre-size matrices: std::vector<std::vector<int>>(rows, std::vector<int>(cols)).
 *     - For numerical heavy lifting, prefer Eigen / xtensor / BLAS.
 *
 * DRY RUN:
 *     mat = {{1,2,3},{4,5,6},{7,8,9}}
 *     transpose -> {{1,4,7},{2,5,8},{3,6,9}}
 *     rotate 90 CW: {{7,4,1},{8,5,2},{9,6,3}}
 *     {{1,2},{3,4}} x {{5,6},{7,8}} = {{19,22},{43,50}}
 *
 * COMPLEXITY:
 *     Transpose      : O(rows*cols)
 *     Rotation       : O(n^2)
 *     Multiplication : O(m*n*k)
 */

#include <iostream>
#include <vector>
#include <algorithm>

using Matrix = std::vector<std::vector<int>>;

void printMatrix(const Matrix& mat) {
    for (const auto& row : mat) {
        std::cout << "[";
        for (std::size_t i = 0; i < row.size(); ++i)
            std::cout << row[i] << (i + 1 < row.size() ? ", " : "");
        std::cout << "]\n";
    }
    std::cout << "\n";
}

void transposeSquare(Matrix& mat) {
    int n = static_cast<int>(mat.size());
    for (int i = 0; i < n; ++i)
        for (int j = i + 1; j < n; ++j)
            std::swap(mat[i][j], mat[j][i]);
}

Matrix transpose(const Matrix& mat) {
    int rows = static_cast<int>(mat.size());
    int cols = static_cast<int>(mat[0].size());
    Matrix result(cols, std::vector<int>(rows, 0));
    for (int i = 0; i < rows; ++i)
        for (int j = 0; j < cols; ++j)
            result[j][i] = mat[i][j];
    return result;
}

void reverseRows(Matrix& mat) {
    for (auto& row : mat)
        std::reverse(row.begin(), row.end());
}

void rotate90CW(Matrix& mat) {
    transposeSquare(mat);
    reverseRows(mat);
}

void rotate90CCW(Matrix& mat) {
    reverseRows(mat);
    transposeSquare(mat);
}

Matrix multiply(const Matrix& A, const Matrix& B) {
    int m = static_cast<int>(A.size());
    int k = static_cast<int>(A[0].size());
    int n = static_cast<int>(B[0].size());
    Matrix C(m, std::vector<int>(n, 0));
    for (int i = 0; i < m; ++i)
        for (int j = 0; j < n; ++j)
            for (int p = 0; p < k; ++p)
                C[i][j] += A[i][p] * B[p][j];
    return C;
}

int main() {
    Matrix mat = {{1,2,3},{4,5,6},{7,8,9}};
    std::cout << "Original:\n"; printMatrix(mat);

    Matrix mat2 = mat;
    transposeSquare(mat2);
    std::cout << "Transposed (square, in-place):\n"; printMatrix(mat2);

    Matrix nonSquare = {{1,2,3},{4,5,6}};
    std::cout << "Non-square (2x3):\n"; printMatrix(nonSquare);
    std::cout << "Transposed (3x2):\n"; printMatrix(transpose(nonSquare));

    Matrix mat3 = mat;
    rotate90CW(mat3);
    std::cout << "Rotated 90° CW:\n"; printMatrix(mat3);
    // Expected: {{7,4,1},{8,5,2},{9,6,3}}

    Matrix A = {{1,2},{3,4}};
    Matrix B = {{5,6},{7,8}};
    std::cout << "A x B:\n"; printMatrix(multiply(A, B));
    // {{19,22},{43,50}}

    return 0;
}

/*
 * NOTES — C++ vs Java:
 *     - std::vector<std::vector<int>> is the natural 2D container.
 *     - std::swap and std::reverse provide the primitive ops.
 *     - For production numeric work, Eigen and xtensor offer high-performance
 *       expression templates and BLAS-backed matmul.
 */
